### ISSUE TITLE
Unicode dir is seen as not a dir on windows only via os.scandir...

### ISSUE THREAD

**Issue by Carreau** [pypy/pypy#4860]
This is quite recent failure on IPython, 

I'm using https://github.com/ipython/ipython/pull/14293 to debug, and already pinged `@mattip`.

It looks like somehow between 7.3.13, and 7.3.14, folder with unicode names are reported as not-a-directory on windows.

```
    def test_wierd_pypy_73713_7314():
        TMP = Path(tempfile.mkdtemp())
        td = Path(tempfile.mkdtemp(dir=TMP))
        p = Path(td / "profile_ünicode2") 
        p.mkdir(parents=True) # p should really be a dir
        assert p.is_dir() # Yep it's s dir.
        l = list(os.scandir(td)) # let's get it via scandir.
>       assert l[0].is_dir(), ("l[0]", l[0], "should be dir is not") 
E       AssertionError: ('l[0]', <DirEntry 'profile_ünicode2'>, 'should be dir is not')
E       assert False
E        +  where False = <bound method DirEntry.is_dir of <DirEntry 'profile_ünicode2'>>()
E        +    where <bound method DirEntry.is_dir of <DirEntry 'profile_ünicode2'>> = <DirEntry 'profile_ünicode2'>.is_dir
```

**Comment by mattip** [comment]
@cpburnz any thoughts about what is wrong? Maybe something in the PR to handle symlinks in pypy + windows + os.scandir is a bit off?

**Comment by mattip** [comment]
Debugging this, the difference comes from [W_DirEntry.check_mode](https://github.com/pypy/pypy/blob/ae536562f5b45ccbd8a0f61ad8622fb7a657d003/pypy/module/posix/interp_scandir.py#L317), which used to simply [return `self.flags`](https://github.com/pypy/pypy/blob/bd1a0d059f5e947f45cffb2c0293049ea2d30557/pypy/module/posix/interp_scandir.py#L325). In the PR to support symlinks, it was changed to call `get_stat_or_lstat`, which chokes on unicode names.

**Comment by mattip** [comment]
mess. The change was done to allow following symlinks, and ends up calling `stat` or `lstat`. It turns out `rposix.stat()` and `rposix.lstat()` are still using the paradigm where a path with non-ascii characters is converted to python2 unicode, and then fed through `CreateFileW`, where an ascii character path uses `CreateFileA`. The call in `check_mode` converts the path to utf8, which is mistakenly fed through as ascii, so the call to `lstat` silently fails and returns `-1`, which is then interpreted as "not a directory"

The correct fix is to continue to use utf8, and to convert the os.stat/os.lstat calls to always use `CreateFileW` by converting the utf8 string to windows wchar.

**Comment by Carreau** [comment]
bleh. i'm both happy IPython test suite caught it, and sorry for putting
you through some pain.

**Comment by mattip** [comment]
well, I went with the incorrect 1 line fix to continue to use unicode via `FileEncoder`. Someday someone may invest some time to clean up windows.

**Comment by mattip** [comment]
@Carreau I don't know if this is worth a new release. Could you skip that test for `sys.implementation.name == 'pypy' and sys.implementation.version < (7.3.16)`?

**Comment by Carreau** [comment]
Yeah, no worries I can do that on my side.

**Comment by cpburnz** [comment]
@mattip I didn't think to test unicode paths. I copied over the path handling from the unused (at the time) Win32 `W_DirEntry.get_stat_or_lstat()` method because it handled it clearly differently than the POSIX methods.

**Comment by mattip** [comment]
@cpburnz OK, thanks. Can you take a look at #4864 and think of any tests I might have missed?

**Comment by cpburnz** [comment]
@mattip `DirEntry.is_dir()` ends up calling `.get_stat()` which always calls `.get_lstat()`, and in turn its use of `rposix_stat.lstat()` is covered by `test_unicode_dir()` (in test_scandir.py). `.get_stat()` only calls `rposix_stat.stat()` when when a symlink is encountered so that use is not covered by `test_unicode_dir()`. To have comprehensive coverage, I think another test is needed: a symlink to a directory with unicode name.

**Comment by mattip** [comment]
Right. So that would have to be in the extra_tests/test_scandir.py tests in order to run after translation, since windows cannot really test scandir symlinks when testing untranslated.

**Comment by mattip** [comment]
Closing as fixed by #4873.
